lib/widgets/action_sheet.dart

```diff
@@ -181,6 +142,264 @@ class MessageActionSheetCancelButton extends StatelessWidget {
+            optionButtons.add(follow);
+          }
+        case UserTopicVisibilityPolicy.unmuted:
+          optionButtons.add(mute);
+          if (supportsFollowingTopics) {
+            optionButtons.add(follow);
+          }
+        case UserTopicVisibilityPolicy.followed:
+          optionButtons.add(mute);
+          if (supportsFollowingTopics) {
+            optionButtons.add(unfollow);
+          }
+        case UserTopicVisibilityPolicy.unknown:
+          // TODO(#1074): This should be unreachable as we keep `unknown` out of
+          //   our data structures.
+          assert(false);
+      }
+    }
+  } else {
+    // Not subscribed to the channel; there is no user topic change to be made.
+  }
+
+  if (optionButtons.isEmpty) {
+    // TODO(a11y): While long press has no effect when we return early without
+    //   bringing up the action sheet, the screen readers do not have a way to
+    //   know that.  However, we may return this early return after we add a
+    //   always-present button.
+    return;
+  }
+
+  _showActionSheet(context, optionButtons: optionButtons);
+}
+
+class UserTopicUpdateButton extends ActionSheetMenuItemButton {
+  const UserTopicUpdateButton({
+    super.key,
+    this.currentVisibilityPolicy,
+    required this.newVisibilityPolicy,
+    required this.narrow,
+    required super.pageContext,
+  });
+
+  final UserTopicVisibilityPolicy? currentVisibilityPolicy;
+  final UserTopicVisibilityPolicy newVisibilityPolicy;
+  final TopicNarrow narrow;
+
+  @override IconData get icon {
+    switch (newVisibilityPolicy) {
+      case UserTopicVisibilityPolicy.none:
+        return ZulipIcons.inherit;
+      case UserTopicVisibilityPolicy.muted:
+        return ZulipIcons.mute;
+      case UserTopicVisibilityPolicy.unmuted:
+        return ZulipIcons.unmute;
+      case UserTopicVisibilityPolicy.followed:
+        return ZulipIcons.follow;
+      case UserTopicVisibilityPolicy.unknown:
+        // TODO(#1074): This should be unreachable as we keep `unknown` out of
+        //   our data structures.
+        assert(false);
+        return ZulipIcons.inherit;
+    }
+  }
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    switch ((currentVisibilityPolicy, newVisibilityPolicy)) {
+      case (UserTopicVisibilityPolicy.muted, UserTopicVisibilityPolicy.none):
+        return zulipLocalizations.actionSheetOptionUnmuteTopic;
+      case (UserTopicVisibilityPolicy.followed, UserTopicVisibilityPolicy.none):
+        return zulipLocalizations.actionSheetOptionUnfollowTopic;
+
+      case (_, UserTopicVisibilityPolicy.muted):
+        return zulipLocalizations.actionSheetOptionMuteTopic;
+      case (_, UserTopicVisibilityPolicy.unmuted):
+        return zulipLocalizations.actionSheetOptionUnmuteTopic;
+      case (_, UserTopicVisibilityPolicy.followed):
+        return zulipLocalizations.actionSheetOptionFollowTopic;
+
+      case (_, UserTopicVisibilityPolicy.none):
+        return '';
```

Comment:
It turned out that this would be better as a separate case and keep the `assert(false)`. Added an explanation below:

```dart
      case (_, UserTopicVisibilityPolicy.none):
        // This is unexpected because `UserTopicVisibilityPolicy.muted` and
        // `UserTopicVisibilityPolicy.followed` (handled in separate `case`'s)
        // are the only expected `currentVisibilityPolicy`
        // when `newVisibilityPolicy` is `UserTopicVisibilityPolicy.none`.
        assert(false);
        return '';
```